python_modules/dagster/dagster/_core/definitions/freshness_checks/sensor.py

```diff
@@ -0,0 +1,161 @@
+from typing import Any, Mapping, Optional, Sequence
+
+import pendulum
+from pydantic.dataclasses import dataclass
+
+from dagster import _check as check
+from dagster._core.definitions.asset_check_spec import AssetCheckKey, AssetCheckSpec
+from dagster._core.definitions.asset_selection import AssetSelection
+from dagster._core.definitions.run_request import RunRequest
+from dagster._serdes.serdes import deserialize_value, serialize_value, whitelist_for_serdes
+from dagster._utils.merger import merge_dicts
+from dagster._utils.schedules import get_latest_completed_cron_tick
+
+from ..asset_checks import AssetChecksDefinition
+from ..decorators import sensor
+from ..sensor_definition import SensorDefinition, SensorEvaluationContext
+from .non_partitioned import NON_PARTITIONED_FRESHNESS_PARAMS_METADATA_KEY
+from .time_window_partitioned import TIME_WINDOW_PARTITIONED_FRESHNESS_PARAMS_METADATA_KEY
+from .utils import (
+    DEFAULT_FRESHNESS_CRON_TIMEZONE,
+    FRESHNESS_CRON_METADATA_KEY,
+    FRESHNESS_CRON_TIMEZONE_METADATA_KEY,
+    MAXIMUM_LAG_METADATA_KEY,
+    ensure_freshness_checks,
+    ensure_no_duplicate_asset_checks,
+)
+
+DEFAULT_FRESHNESS_SENSOR_NAME = "freshness_checks_sensor"
+
+
+@whitelist_for_serdes
+@dataclass(frozen=True)
+class FreshnessCheckSensorCursor:
+    """The cursor for the freshness check sensor."""
+
+    evaluation_timestamps: Mapping[str, float]
+
+    def get_last_evaluation_timestamp(self, key: AssetCheckKey) -> Optional[float]:
+        return self.evaluation_timestamps.get(key.to_user_string())
+
+    @staticmethod
+    def initialize():
+        return FreshnessCheckSensorCursor(evaluation_timestamps={})
+
+    def with_updated_evaluations(
+        self, keys: Sequence[AssetCheckKey], evaluation_timestamp: float
+    ) -> "FreshnessCheckSensorCursor":
+        return FreshnessCheckSensorCursor(
+            evaluation_timestamps=merge_dicts(
+                self.evaluation_timestamps,
+                {key.to_user_string(): evaluation_timestamp for key in keys},
+            )
+        )
+
+
+def build_freshness_check_sensor(
+    *,
+    freshness_checks: Sequence[AssetChecksDefinition],
+    interval_seconds: Optional[int] = None,
+    name: str = DEFAULT_FRESHNESS_SENSOR_NAME,
+) -> SensorDefinition:
+    """Builds a sensor which kicks off evaluation of freshness checks.
```

Comment:
Worth calling out that it looks at the parameters on the freshness checks to determine how frequently to kick them off.